File: apk/app/src/main/java/app/momentra/ui/theme/DesignTokens.kt
package app.momentra.ui.theme

import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.graphics.Color
import java.text.NumberFormat
import java.util.Locale

/**
 * Unified entry point for design tokens (parity with iOS `DesignTokens`).
 */
object DesignTokens {
    val base = MomentraBase
    val urgency = MomentraUrgency
    val personal = PersonalColors
    val group = GroupColors
    val business = BusinessColors
    val circle = CircleColors
    val splash = SplashColors
    val avatar = AvatarColors
    val semantic = MomentraSemantic

    val spacing = MomentraSpacing
    val radius = MomentraRadius
    val sizing = MomentraSizing
    val type = MomentraType
    val anim = MomentraAnim
    val icon = MomentraIconSizing

    /** Same role as iOS `DesignTokens.theme(for:)`. */
    fun theme(forContext: MomentraContext): MomentraContextTheme =
        MomentraContextThemes.themeFor(forContext)

    enum class UrgencyLevel { Clear, Medium, High }
    enum class ActionRole { Primary, RemindOverdue }

    data class ActionStyle(
        val solid: Color,
        val solidAlt: Color,
        val gradientStart: Color,
        val gradientEnd: Color,
        val text: Color,
    )

    fun urgencyLevelForStatus(status: String?): UrgencyLevel {
        val normalized = status?.trim()?.lowercase().orEmpty()
        return when {
            normalized.contains("overdue") || normalized.contains("high") || normalized.contains("late") -> UrgencyLevel.High
            normalized.contains("medium") || normalized.contains("warning") || normalized.contains("due_soon") -> UrgencyLevel.Medium
            else -> UrgencyLevel.Clear
        }
    }

    fun actionStyle(
        context: MomentraContext,
        status: String?,
        role: ActionRole = ActionRole.Primary,
    ): ActionStyle {
        val level = urgencyLevelForStatus(status)
        val forceUrgencyCta = role == ActionRole.RemindOverdue || level != UrgencyLevel.Clear
        return if (forceUrgencyCta) {
            ActionStyle(
                solid = urgency.cta,
                solidAlt = urgency.ctaEnd,
                gradientStart = urgency.cta,
                gradientEnd = urgency.ctaEnd,
                text = semantic.ctaText,
            )
        } else {
            val theme = theme(forContext = context)
            ActionStyle(
                solid = theme.accent,
                solidAlt = theme.accentEnd,
                gradientStart = theme.accent,
                gradientEnd = theme.accentEnd,
                text = semantic.ctaText,
            )
        }
    }

    /** INR display — parity with iOS `DesignTokens.formatInr`. */
    fun formatInr(amount: Double): String {
        val nf = NumberFormat.getCurrencyInstance(Locale("en", "IN"))
        return nf.format(amount)
    }

    /** Compact INR for hero cards — parity with iOS `DesignTokens.formatInrDisplay`. */
    fun formatInrDisplay(amount: Double): String {
        val absAmount = kotlin.math.abs(amount)
        return when {
            absAmount >= 10_000_000 ->
                String.format(Locale.US, "₹%.1fCr", absAmount / 10_000_000)
            absAmount >= 100_000 ->
                String.format(Locale.US, "₹%.1fL", absAmount / 100_000)
            absAmount >= 1_000 ->
                String.format(Locale.US, "₹%.1fk", absAmount / 1_000)
            else -> {
                val nf = NumberFormat.getNumberInstance(Locale("en", "IN"))
                nf.maximumFractionDigits = 0
                "₹${nf.format(absAmount)}"
            }
        }
    }
}

object MomentraIconSizing {
    val xs: Dp = 12.dp
    val sm: Dp = 16.dp
    val md: Dp = 20.dp
    val lg: Dp = 24.dp
}

File: momentra/momentra/DesignTokens.swift
//
//  DesignTokens.swift
//  momentra
//
//  Created by santosh on 19/04/26.
//

import SwiftUI

/// Unified namespace for all design tokens in Momentra.
/// This wraps and re-exports color, spacing, typography, radius, sizing, and animation tokens.
/// Prefer using `DesignTokens` as your entry point in views.
enum DesignTokens {

    // MARK: - Contexts & Theme
    static let context = MomentraContext.self
    
    /// Get theme for a specific context
    static func theme(for context: MomentraContext) -> ContextTheme {
        ContextTheme.theme(for: context)
    }

    enum UrgencyLevel {
        case clear
        case medium
        case high
    }

    enum ActionRole {
        case primary
        case remindOverdue
    }

    struct ActionStyle {
        let solid: Color
        let solidAlt: Color
        let gradientStart: Color
        let gradientEnd: Color
        let text: Color
    }

    static func urgencyLevel(for status: String?) -> UrgencyLevel {
        let normalized = status?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() ?? ""
        if normalized.contains("overdue") || normalized.contains("high") || normalized.contains("late") {
            return .high
        }
        if normalized.contains("medium") || normalized.contains("warning") || normalized.contains("due_soon") {
            return .medium
        }
        return .clear
    }

    static func actionStyle(
        context: MomentraContext,
        status: String?,
        role: ActionRole = .primary
    ) -> ActionStyle {
        let level = urgencyLevel(for: status)
        let useUrgencyCta = role == .remindOverdue || level != .clear
        if useUrgencyCta {
            return ActionStyle(
                solid: urgency.cta,
                solidAlt: urgency.ctaEnd,
                gradientStart: urgency.cta,
                gradientEnd: urgency.ctaEnd,
                text: semantic.ctaText
            )
        }

        let themeTokens = theme(for: context)
        return ActionStyle(
            solid: themeTokens.accent,
            solidAlt: themeTokens.accentEnd,
            gradientStart: themeTokens.accent,
            gradientEnd: themeTokens.accentEnd,
            text: semantic.ctaText
        )
    }

    // MARK: - Base Colors
    static let base = MomentraBase.self

    // MARK: - Urgency Colors
    static let urgency = MomentraUrgency.self

    // MARK: - Personal Context Colors
    static let personal = PersonalColors.self

    // MARK: - Group Context Colors
    static let group = GroupColors.self

    // MARK: - Business Context Colors
    static let business = BusinessColors.self

    // MARK: - Circle Context Colors
    static let circle = CircleColors.self
    
    // MARK: - Splash colors
    static let splash = SplashColors.self

    // MARK: - Typography
    static let type = MomentraType.self

    // MARK: - Spacing
    static let spacing = MomentraSpacing.self

    // MARK: - Radius
    static let radius = MomentraRadius.self

    // MARK: - Sizing
    static let sizing = MomentraSizing.self

    // MARK: - Animation
    static let anim = MomentraAnim.self

    // MARK: - Avatar Colors
    static let avatar = AvatarColors.self
    
    // MARK: - Semantic roles
    static let semantic = MomentraSemantic.self

    // MARK: - INR Formatter
    static func formatInr(_ amount: Double) -> String {
        MomentraFormatters.formatInr(amount)
    }
    
    static func formatInrDisplay(_ amount: Double) -> String {
        MomentraFormatters.formatInrDisplay(amount)
    }
}

// Usage Examples:
//
// let theme = DesignTokens.theme(for: .personal)
// Text("Amount").foregroundColor(DesignTokens.personal.text)
// .font(DesignTokens.type.display)
//
// .padding(.horizontal, DesignTokens.spacing.screenH)
// .cornerRadius(DesignTokens.radius.cardSm)
